lightning/src/offers/signer.rs

```diff
@@ -0,0 +1,130 @@
+
+/// Message metadata which possibly is derived from [`MetadataMaterial`] such that it can be
+/// verified.
+#[derive(Clone)]
+pub(super) enum Metadata {
+	/// Metadata as parsed, supplied by the user, or derived.
+	Bytes(Vec<u8>),
+
+	/// Metadata to be derived from message contents and given material.
+	Derived(MetadataMaterial),
+
+	/// Metadata and signing pubkey to be derived from message contents and given material.
+	DerivedSigningPubkey(MetadataMaterial),
+}
+
+impl Metadata {
+	pub fn as_bytes(&self) -> Option<&Vec<u8>> {
+		match self {
+			Metadata::Bytes(bytes) => Some(bytes),
+			Metadata::Derived(_) => None,
+			Metadata::DerivedSigningPubkey(_) => None,
+		}
+	}
+
+	pub fn has_derivation_material(&self) -> bool {
+		match self {
+			Metadata::Bytes(_) => false,
+			Metadata::Derived(_) => true,
+			Metadata::DerivedSigningPubkey(_) => true,
+		}
+	}
+
+	pub fn without_keys(self) -> Self {
+		match self {
+			Metadata::Bytes(_) => self,
+			Metadata::Derived(_) => self,
+			Metadata::DerivedSigningPubkey(material) => Metadata::Derived(material),
+		}
+	}
+
+	pub fn derive_from<W: Writeable>(self, tlv_stream: W) -> (Self, Option<KeyPair>) {
+		match self {
+			Metadata::Bytes(_) => (self, None),
+			Metadata::Derived(mut metadata_material) => {
+				tlv_stream.write(&mut metadata_material.hmac).unwrap();
+				(Metadata::Bytes(metadata_material.derive_metadata()), None)
+			},
+			Metadata::DerivedSigningPubkey(mut metadata_material) => {
+				tlv_stream.write(&mut metadata_material.hmac).unwrap();
+				let (metadata, keys) = metadata_material.derive_metadata_and_keys();
+				(Metadata::Bytes(metadata), Some(keys))
+			},
+		}
+	}
+}
+
+impl fmt::Debug for Metadata {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			Metadata::Bytes(bytes) => bytes.fmt(f),
+			Metadata::Derived(_) => f.write_str("Derived"),
+			Metadata::DerivedSigningPubkey(_) => f.write_str("DerivedSigningPubkey"),
+		}
+	}
+}
+
+impl PartialEq for Metadata {
+	fn eq(&self, other: &Self) -> bool {
+		match self {
+			Metadata::Bytes(bytes) => if let Metadata::Bytes(other_bytes) = other {
+				bytes == other_bytes
+			} else {
+				false
+			},
+			Metadata::Derived(_) => false,
+			Metadata::DerivedSigningPubkey(_) => false,
+		}
+	}
+}
+
+/// Material used to create metadata for a message.
+#[derive(Clone)]
+pub(super) struct MetadataMaterial {
+	nonce: EncryptedNonce,
+	hmac: HmacEngine<Sha256>,
+}
+
+impl MetadataMaterial {
+	pub fn new(nonce: Nonce, expanded_key: &ExpandedKey, iv_bytes: &[u8; IV_LEN]) -> Self {
+		Self {
+			nonce: expanded_key.encrypt_nonce_for_offer(nonce, iv_bytes),
+			hmac: expanded_key.hmac_for_offer(nonce),
+		}
+	}
+
+	fn derive_metadata(self) -> Vec<u8> {
+		let mut bytes = self.nonce.as_slice().to_vec();
+		bytes.extend_from_slice(&Hmac::from_engine(self.hmac).into_inner());
+		bytes
+	}
+
+	fn derive_metadata_and_keys(self) -> (Vec<u8>, KeyPair) {
+		let secp_ctx = Secp256k1::new();
```

Comment:
Would you care if a context is cloned for this? Ideally, it is passed when the builder is created and carried along either in the applicable `Metadata` variants or directly in `MetadataMaterial`. Otherwise, if a reference is kept, the lifetime would trickle up to the top-level messages (i.e., `Offer`, `Refund`, etc.).

An alternative would be to pass it in to the `build` calls. But then either some will take a context that is never used or we further make separate `build` methods for the derived variants. I kinda wanted to avoid passing to `build` though because that means the `ChannelManager` utilities for creating builders in the next PR won't return fully usable builders; the caller would need to fetch the context when eventually calling `build`.